### DISCUSSION TITLE
Flattening multiple subsequent calls to the same method in to one call

### DISCUSSION THREAD

**Discussion by evanchooly** [openrewrite/rewrite#4190] [Q&A] [answered]
What I'm trying to do is perhaps best related by [show](https://github.com/MorphiaOrg/morphia/blob/f19fd0d5c864308286da5e564710d87e3709cee8/rewrite/src/test/java/dev/morphia/rewrite/recipes/test/PipelineRewriteTest.java#L40) rather than tell.  In essence, I want to pivot to one varargs method (name `pipeline()`) that takes a `Stage...` vs the current named method per stage approach.  At Devnexus, Jonathan's advice on my recipe was to first convert each named stage method to a `pipeline()` call.  Once that is done, condense those back to back calls to just one by collecting all the arguments from of the invocations, reduce it to just one call, and pass in all those arguments.

I have the first part [done](https://github.com/MorphiaOrg/morphia/blob/f19fd0d5c864308286da5e564710d87e3709cee8/rewrite/src/main/java/dev/morphia/rewrite/recipes/PipelineRewrite.java#L84) as you can see by the test above.  I'm successfully rewriting the stage methods in to `pipeline()` calls but I'm at a loss as to how collect all the `pipeline()` calls such that I can condense them in to one.  Any advice for me?

**Comment by timtebeek** [comment, marked as answer]
Hi @evanchooly ; Our Slack #questions channel might get a few more eyeballs to help you out, especially as I'm traveling today and tomorrow still, so I can only provide quick feedback here.

Did you already try to visit any `.pipeline` method invocation, and from there inspect the chain preceding it by looking at `getSelect()`? That might help you gather up all the stages, which you can then replace with a single `.pipeline` call by using a JavaTemplate with coordinates `.replace()` instead of `.replaceMethod()`.

It's not a perfect match, but [this earlier work](https://github.com/openrewrite/rewrite-testing-frameworks/blob/main/src/main/java/org/openrewrite/java/testing/assertj/SimplifyChainedAssertJAssertion.java) came to mind as a sort of similar use case that needs preceding elements to make the change.
